File: app/src/main/java/ru/alemak/studentapp/data/repository/NewsRepository.kt
package ru.alemak.studentapp.data.repository

import javax.inject.Inject
import javax.inject.Singleton
import ru.alemak.studentapp.data.local.NewsCacheEntity
import ru.alemak.studentapp.data.local.NewsDao
import ru.alemak.studentapp.data.model.NewsItem
import ru.alemak.studentapp.data.remote.ScheduleApi

data class NewsLoadResult(
    val news: List<NewsItem>,
    /** true only if we failed to reach API and fell back to local Room cache */
    val fromCache: Boolean,
    val updatedAtMillis: Long = 0L,
)

@Singleton
class NewsRepository @Inject constructor(
    private val api: ScheduleApi,
    private val newsDao: NewsDao,
) {
    suspend fun getNewsFromCacheOnly(limit: Int = 15): NewsLoadResult {
        val cachedEntities = newsDao.getAll().take(limit)
        val cached = cachedEntities.map { it.toDomain() }
        val updatedAt = cachedEntities.maxOfOrNull { it.updatedAt } ?: 0L
        return NewsLoadResult(
            news = cached,
            fromCache = cached.isNotEmpty(),
            updatedAtMillis = updatedAt,
        )
    }

    suspend fun getNews(limit: Int = 15, force: Boolean = false): NewsLoadResult {
        // Try network; on VPN failure fall back to Room instantly usable cache.
        val response = runCatching { api.getNews(limit = limit, force = force) }.getOrNull()
        val remote = response?.news?.map { it.toDomain() }

        if (remote != null && remote.isNotEmpty()) {
            val serverTs = response.updatedAt?.let { (it * 1000.0).toLong() }?.takeIf { it > 0 }
            val now = serverTs ?: System.currentTimeMillis()
            newsDao.clear()
            newsDao.upsertAll(
                remote.mapIndexed { index, item ->
                    NewsCacheEntity.fromDomain(item, index).copy(updatedAt = now)
                },
            )
            return NewsLoadResult(remote, fromCache = false, updatedAtMillis = now)
        }

        val cachedEntities = newsDao.getAll()
        val cached = cachedEntities.map { it.toDomain() }
        val updatedAt = cachedEntities.maxOfOrNull { it.updatedAt } ?: 0L
        // Empty remote (or fail) → keep previous cache rather than blank home
        if (remote != null && remote.isEmpty() && cached.isNotEmpty()) {
            return NewsLoadResult(cached, fromCache = true, updatedAtMillis = updatedAt)
        }
        return NewsLoadResult(
            news = cached,
            fromCache = cached.isNotEmpty(),
            updatedAtMillis = updatedAt,
        )
    }
}

File: ios/StudentKGTU_IOS/Services/NewsRepository.swift
import Foundation

actor NewsRepository {
    static let shared = NewsRepository()

    func getNews(limit: Int = 10) async -> (news: [NewsItem], fromCache: Bool, updatedAt: Double) {
        let key = "news_\(limit)"

        // Offline: cache only, no long network wait
        if !NetworkMonitor.shared.isOnline {
            let cached = JSONCache.load([NewsItem].self, key: key) ?? []
            return (cached, true, JSONCache.meta(key: key))
        }

        do {
            let remote = try await APIClient.shared.news(limit: limit)
            let now = Date().timeIntervalSince1970 * 1000
            JSONCache.save(remote, key: key)
            JSONCache.saveMeta(key: key, updatedAt: now)
            return (remote, false, now)
        } catch {
            let cached = JSONCache.load([NewsItem].self, key: key) ?? []
            return (cached, true, JSONCache.meta(key: key))
        }
    }

    /// Instant disk read without network.
    nonisolated func getNewsFromCacheOnly(limit: Int = 10) -> (news: [NewsItem], updatedAt: Double) {
        let key = "news_\(limit)"
        let cached = JSONCache.load([NewsItem].self, key: key) ?? []
        return (cached, JSONCache.meta(key: key))
    }
}
